src/material-experimental/mdc-chips/chip-cell.ts

```diff
@@ -6,14 +6,271 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {Directive} from '@angular/core';
+import {coerceBooleanProperty} from '@angular/cdk/coercion';
+import {Platform} from '@angular/cdk/platform';
+import {AfterContentInit, AfterViewInit, Component, ChangeDetectorRef, ContentChild, Directive, ElementRef, EventEmitter, Input, NgZone, OnDestroy, Output, ViewChild} from '@angular/core';
+import {RippleRenderer, RippleTarget, ThemePalette} from '@angular/material/core';
+import {MDCChipAdapter, MDCChipFoundation} from '@material/chips';
+import {Subject} from 'rxjs';
+import {takeUntil} from 'rxjs/operators';
 
+let uid = 0;
+
+/** Represents an event fired on an individual `mat-chip`. */
+export interface MatChipCellEvent {
+  /** The chip the event was fired on. */
+  chip: MatChipCell;
+}
+
+/**
+ * Directive to add CSS classes to chip leading icon.
+ * @docs-private
+ */
 @Directive({
-  selector: 'mat-chip-cell',
+  selector: 'mat-chip-avatar, [matChipAvatar]',
+  host: {
+    'class': 'mat-chip-avatar mdc-chip__icon mdc-chip__icon--leading',
+  }
+})
+export class MatChipAvatar {
+  constructor(private _changeDetectorRef: ChangeDetectorRef,
+              private _elementRef: ElementRef) {}
+
+  /** Sets whether the given CSS class should be applied to the leading icon. */
+  setClass(cssClass: string, active: boolean) {
+    const element = this._elementRef.nativeElement;
+    active ? element.addClass(cssClass) : element.removeClass(cssClass);
+    this._changeDetectorRef.markForCheck();
+  }
+}
+
+/**
+ * Directive to add CSS class to chip trailing icon and notify parent chip
+ * about trailing icon interactions.
+ *
+ * If matChipRemove is used to add this directive, the parent chip will be
+ * removed when the trailing icon is clicked.
+ *
+ * @docs-private
+z */
+@Directive({
+  selector: 'mat-chip-trailing-icon, [matChipTrailingIcon], [matChipRemove]',
+  host: {
+    'class': 'mat-chip-trailing-icon mdc-chip__icon mdc-chip__icon--trailing',
+    'tabindex': '0',
+    'role': 'button',
+    '(click)': 'interaction.emit($event)',
+    '(keydown)': 'interaction.emit($event)',
+  }
+})
+export class MatChipTrailingIcon {
+  @Output() interaction = new EventEmitter<MouseEvent | KeyboardEvent>();
+  shouldRemove!: boolean;
+
+  constructor(_elementRef: ElementRef) {
+    this.shouldRemove = _elementRef.nativeElement.getAttribute('matChipRemove') !== null;
+  }
+}
+
+/**
+ * Material design styled Chip component. Used inside the MatChipGrid component.
+ */
+@Component({
+  selector: 'mat-chip-cell, mat-basic-chip-cell',
+  templateUrl: 'chip-cell.html',
+  styleUrls: ['chips.css'],
   host: {
     'class': 'mat-mdc-chip-cell',
+    '[id]': 'id',
+    '[class.mat-chip-disabled]': 'disabled',
+    '[class.mat-chip-selected]': 'selected',
   },
 })
-export class MatChipCell {
-  // TODO: set up MDC foundation class.
+export class MatChipCell implements AfterContentInit, AfterViewInit, OnDestroy {
```

Comment:
One change I would want to make at a high level is to have the "grid/cell" pattern for the chips be a subclass or some other kind of variant of a more general "MatChipSet" / "MatChip" base. We do this with our list component now, where we have these different variants: `mat-list`, `mat-nav-list`, `mat-selection-list`, etc. Each has a different interaction pattern, but they largely share the same implementation. 

I chatted w/ @mmalerba about it and we think we could just use basic inheritance for this. The downside is that the child classes have to duplicate templates and host bindings since Angular (before ivy) doesn't propagate those downward.

[The other approach we considered would be to make common components that matched all of the selectors (`mat-chip-grid`, `mat-chip-listbox`, etc.) and then have an extra directive for each that added their special behavior for their specific selector.]

Reply:
I remember we discussed this, but was fuzzy on how to go about implementing it. Thanks for providing some more detail! I attempted to refactor the code so that there are MatChipOption and MatChipCell components that extend the MatChip base component, and same for MatChipGrid/MatChipList extending MatChipSet. I'll push the branch so you can see. It's still messy and things aren't wired up yet, but is this generally the structure that you're looking for? I'm not super familiar with inheritance in angular. I did take a look at the list components that you mentioned, and I see that they don't extend one base component. In what way do they share an implementation?